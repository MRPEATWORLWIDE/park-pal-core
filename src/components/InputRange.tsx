import { useEffect, useRef, useState } from 'react'
import styles from './InputRange.module.scss'

type Props = {
  defaultValue?: [number, number]
  min?: number
  max?: number
  step?: number
}

export const InputRange: React.FC<Props> = ({ min = 0, max = 100, defaultValue = [min, max], step = 1 }) => {
  const progressRef = useRef<HTMLDivElement>(null)
  const [minValue, setMinValue] = useState<number>(defaultValue[0] < min ? min : defaultValue[0])
  const [maxValue, setMaxValue] = useState<number>(defaultValue[1] > max ? max : defaultValue[1])

  const gap = 0

  useEffect(() => {
    if (progressRef.current) {
      progressRef.current.style.left = ((minValue / max) * 100) + "%"
      progressRef.current.style.right = 100 - (maxValue / max) * 100 + "%"
    }
  }, [minValue, maxValue, max])

  const handleMinValueChange: React.ChangeEventHandler<HTMLInputElement> = (event) => {
    const minVal = event.currentTarget.valueAsNumber
    const maxVal = maxValue

    if ((maxVal - minVal) < gap) {
      setMinValue(maxVal - gap)
    } else {
      setMinValue(minVal)
    }
  }

  const handleMaxValueChange: React.ChangeEventHandler<HTMLInputElement> = (event) => {
    const minVal = minValue
    const maxVal = event.currentTarget.valueAsNumber

    if ((maxVal - minVal) < gap) {
      setMaxValue(minVal + gap)
    } else {
      setMaxValue(maxVal)
    }
  }

  return (
    <div>
      <div data-testid='number-input-container' className={styles.numberInput}>
        <div className='field'>
          <span>Min</span>
          <input data-testid='number-input-1' onChange={handleMinValueChange} type='number' min={min} value={minValue} step={step} />
        </div>
        <div className='separator'>-</div>
        <div className='field'>
          <span>Max</span>
          <input data-testid='number-input-2' onChange={handleMaxValueChange} type='number' max={max} value={maxValue} step={step} />
        </div>
      </div>
      <div className={styles.slider}>
        <div data-testid='progress' className='progress' ref={progressRef}></div>
      </div>
      <div data-testid='range-input' className={styles.rangeInput}>
        <input data-testid='range-input-1' onChange={handleMinValueChange} type='range' min={min} max={max} value={minValue} step={step} />
        <input data-testid='range-input-2' onChange={handleMaxValueChange} type='range' min={min} max={max} value={maxValue} step={step} />
      </div>
    </div>
  )
}