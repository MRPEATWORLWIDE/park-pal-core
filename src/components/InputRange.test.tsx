import { fireEvent, render } from '@testing-library/react'
import { act } from 'react-dom/test-utils'
import { InputRange } from './InputRange'


test('should match the snapshot', () => {
  const { container } = render(<InputRange />)

  expect(container).toMatchSnapshot()
})

test('should render with default properties', () => {
  const { getByTestId } = render(<InputRange />)

  expect(getByTestId('number-input-1')).toHaveValue(0)
  expect(getByTestId('number-input-2')).toHaveValue(100)
  expect(getByTestId('range-input-1')).toHaveValue('0')
  expect(getByTestId('range-input-2')).toHaveValue('100')
  expect(getByTestId('progress')).toHaveStyle({
    left: '0%',
    right: '0%'
  })
})

test('should render changing the default values', () => {
  const { getByTestId } = render(<InputRange defaultValue={[10, 90]} />)

  expect(getByTestId('number-input-1')).toHaveValue(10)
  expect(getByTestId('number-input-2')).toHaveValue(90)
  expect(getByTestId('range-input-1')).toHaveValue('10')
  expect(getByTestId('range-input-2')).toHaveValue('90')
  expect(getByTestId('progress')).toHaveStyle({
    left: '10%',
    right: '10%'
  })
})

test('should change the lower defaultValue when "min" is set', () => {
  const { getByTestId } = render(<InputRange min={-10} />)

  expect(getByTestId('number-input-1')).toHaveValue(-10)
  expect(getByTestId('number-input-2')).toHaveValue(100)
  expect(getByTestId('range-input-1')).toHaveValue('-10')
  expect(getByTestId('range-input-2')).toHaveValue('100')
  expect(getByTestId('progress')).toHaveStyle({
    left: '0%',
    right: '0%'
  })
})

test('should change the upper defaultValue when "max" is set', () => {
  const { getByTestId } = render(<InputRange max={250} />)

  expect(getByTestId('number-input-1')).toHaveValue(0)
  expect(getByTestId('number-input-2')).toHaveValue(250)
  expect(getByTestId('range-input-1')).toHaveValue('0')
  expect(getByTestId('range-input-2')).toHaveValue('250')
  expect(getByTestId('progress')).toHaveStyle({
    left: '0%',
    right: '0%'
  })
})

test('should change the provided value if lower than "min"', () => {
  const { getByTestId } = render(<InputRange defaultValue={[10, 90]} min={50} />)

  expect(getByTestId('number-input-1')).toHaveValue(50)
  expect(getByTestId('number-input-2')).toHaveValue(90)
  expect(getByTestId('range-input-1')).toHaveValue('50')
  expect(getByTestId('range-input-2')).toHaveValue('90')
  expect(getByTestId('progress')).toHaveStyle({
    left: '0%',
    right: '20%'
  })
})

test('should change the provided value if greater than "max"', () => {
  const { getByTestId } = render(<InputRange defaultValue={[10, 90]} max={50} />)

  expect(getByTestId('number-input-1')).toHaveValue(10)
  expect(getByTestId('number-input-2')).toHaveValue(50)
  expect(getByTestId('range-input-1')).toHaveValue('10')
  expect(getByTestId('range-input-2')).toHaveValue('50')
  expect(getByTestId('progress')).toHaveStyle({
    left: '20%',
    right: '0%'
  })
})

test('should change the value when user slide without violating min-max rules', () => {
  const { getByTestId } = render(<InputRange defaultValue={[10, 90]} min={10} max={90} />)

  const rangeInput1 = getByTestId('range-input-1')
  const rangeInput2 = getByTestId('range-input-2')

  act(() => {
    fireEvent.change(rangeInput1, { target: { value: '0' } })
  })

  expect(rangeInput1).toHaveValue('10')
  expect(rangeInput2).toHaveValue('90')

  act(() => {
    fireEvent.change(rangeInput1, { target: { value: '50' } })
  })

  expect(rangeInput1).toHaveValue('50')
  expect(rangeInput2).toHaveValue('90')

  act(() => {
    fireEvent.change(rangeInput2, { target: { value: '100' } })
  })

  expect(rangeInput1).toHaveValue('50')
  expect(rangeInput2).toHaveValue('90')

  act(() => {
    fireEvent.change(rangeInput2, { target: { value: '80' } })
  })

  expect(rangeInput1).toHaveValue('50')
  expect(rangeInput2).toHaveValue('80')
})

test('lower value cannot be greater than upper value', () => {
  const { getByTestId } = render(<InputRange defaultValue={[10, 50]} min={10} max={90} />)

  const rangeInput1 = getByTestId('range-input-1')
  const rangeInput2 = getByTestId('range-input-2')

  act(() => {
    fireEvent.change(rangeInput1, { target: { value: '60' } })
  })

  expect(rangeInput1).toHaveValue('50')
  expect(rangeInput2).toHaveValue('50')

  act(() => {
    fireEvent.change(rangeInput1, { target: { value: '20' } })
  })

  act(() => {
    fireEvent.change(rangeInput2, { target: { value: '10' } })
  })

  expect(rangeInput1).toHaveValue('20')
  expect(rangeInput2).toHaveValue('20')
})

test('should render the progress bar correctly', () => {
  const { getByTestId } = render(<InputRange min={9} max={205} />)

  const rangeInput1 = getByTestId('range-input-1')
  const rangeInput2 = getByTestId('range-input-2')
  const progress = getByTestId('progress')

  expect(rangeInput1).toHaveValue('9')
  expect(rangeInput2).toHaveValue('205')
  expect(progress).toHaveStyle({
    left: '0%',
    right: '0%'
  })
})
