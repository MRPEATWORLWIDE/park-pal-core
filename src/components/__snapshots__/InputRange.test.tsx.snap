// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<div>
  <div>
    <div
      class="numberInput"
      data-testid="number-input-container"
    >
      <div
        class="field"
      >
        <span>
          Min
        </span>
        <input
          data-testid="number-input-1"
          min="0"
          step="1"
          type="number"
          value="0"
        />
      </div>
      <div
        class="separator"
      >
        -
      </div>
      <div
        class="field"
      >
        <span>
          Max
        </span>
        <input
          data-testid="number-input-2"
          max="100"
          step="1"
          type="number"
          value="100"
        />
      </div>
    </div>
    <div
      class="slider"
    >
      <div
        class="progress"
        data-testid="progress"
        style="left: 0%; right: 0%;"
      />
    </div>
    <div
      class="rangeInput"
      data-testid="range-input"
    >
      <input
        data-testid="range-input-1"
        max="100"
        min="0"
        step="1"
        type="range"
        value="0"
      />
      <input
        data-testid="range-input-2"
        max="100"
        min="0"
        step="1"
        type="range"
        value="100"
      />
    </div>
  </div>
</div>
`;
